recipes/CoVoST/AST/train_w2v2_llama.py

```diff
@@ -0,0 +1,550 @@
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
+        tokens_prompt_translation, tokens_prompt_translation_len = (
+            batch.tokens_prompt_translation
+        )  # Includes prompt and translation
+        prompt_len = batch.prompt_len
+
+        # Turn padding in the speech to zero. We need to do this in case of leaks, because LLAMA padding is using int of value 120k+ which will corrupt A LOT the signal in case of leak.
+        audio_len = wavs.shape[1]
+        abs_len = torch.round(wav_lens * audio_len)
+        audio_attn_mask = length_to_mask(abs_len)
+        wavs = wavs * audio_attn_mask
+
+        # Add waveform augmentation if specified.
+        if (
+            stage == sb.Stage.TRAIN
+            and hasattr(self.hparams, "wav_augment")
+            and self.optimizer_step > self.hparams.augment_warmup
+        ):
+            wavs, wav_lens = self.hparams.wav_augment(wavs, wav_lens)
+
+        wavs = self.hparams.normalize(wavs, wav_lens)
+
+        # Forward Speech Modules
+        feats = self.modules.wav2vec2(wavs, wav_lens)
+        down_feats = self.modules.feat_downsampler(feats)
+        down_feats_proj = self.modules.proj(down_feats)
+
+        # Format input for LLM; [ audio emb ] + [ prompt emb ] + [ translation emb ]
+        # First get relevant lengths
+        audio_len = down_feats_proj.shape[1]
+        text_len = tokens_prompt_translation.shape[1]
+        audio_prompt_len = (audio_len + prompt_len)[0]
+
+        # Then for the full embedding prompt sequence
+        if hasattr(self.modules.lora_llm, "module"):
+            embeddings = (
+                self.modules.lora_llm.module.model.get_input_embeddings()
+            )
+        else:
+            embeddings = self.modules.lora_llm.model.get_input_embeddings()
+
+        inputs_embeds = torch.cat(
+            (down_feats_proj, embeddings(tokens_prompt_translation)), dim=1
+        )
+
+        # Prepare attn_mask for audio and text and combine them. This is not streaming compatible. For HF to work, masked frames should be 0.
+
+        text_abs_len = torch.round(tokens_prompt_translation_len * text_len)
+        abs_len = torch.round(wav_lens * audio_len)
+        audio_attn_mask = length_to_mask(abs_len)
+        text_attn_mask = length_to_mask(text_abs_len)
+        attn_mask = torch.cat([audio_attn_mask, text_attn_mask], dim=-1)
+
+        # LLM forward
+        llm_logits = self.modules.lora_llm(
+            inputs_embeds=inputs_embeds, attention_mask=attn_mask
+        ).logits
+
+        # output layer for seq2seq log-probabilities
+        p_seq = self.hparams.log_softmax(llm_logits)
+
+        if hasattr(self.modules.lora_llm, "module"):
+            gen_func = self.modules.lora_llm.module.model.generate
+        else:
+            gen_func = self.modules.lora_llm.model.generate
```

Comment:
same as before this is so ugly xD

Reply:
Yes, this is absurdly ugly. I have strictly no idea how we can fix it despite spending a significant amount of time on the problem. We would have to rethink entirely our DDP structure and even then, not sure that we can do something about that.